// @azure/cosmos/lib/request/LocationRouting.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = require("util");
class LocationRouting {
    get ignorePreferredLocation() {
        return this.pIgnorePreferredLocation;
    }
    get locationIndexToRoute() {
        return this.pLocationIndexToRoute;
    }
    get locationEndpointToRoute() {
        return this.pLocationEndpointToRoute;
    }
    routeToLocation(endpointOrIndex, ignorePreferredLocation) {
        if (arguments.length === 2 && util_1.isNumber(endpointOrIndex)) {
            this.pLocationIndexToRoute = endpointOrIndex;
            this.pIgnorePreferredLocation = ignorePreferredLocation;
            this.pLocationEndpointToRoute = undefined;
        }
        else if (arguments.length === 1 && util_1.isString(endpointOrIndex)) {
            this.pLocationEndpointToRoute = endpointOrIndex;
            this.pLocationIndexToRoute = undefined;
            this.pIgnorePreferredLocation = undefined;
        }
        else {
            throw new Error("Invalid arguments passed to routeToLocation");
        }
    }
    clearRouteToLocation() {
        this.pLocationEndpointToRoute = undefined;
        this.pLocationIndexToRoute = undefined;
        this.pIgnorePreferredLocation = undefined;
    }
}
exports.LocationRouting = LocationRouting;
